python/tvm/tir/transform/transform.py

```diff
@@ -310,6 +310,15 @@ def BF16TypeLowering():
     """
     return _ffi_api.BF16TypeLowering()  # type: ignore
 
+def CommonSubexprElim(enable_cse_tir: bool = True):
```

Comment:
There is `disabled_pass` option, so you don't need `enable_cse_tir` flag. 

See https://github.com/apache/tvm/blob/52039c910aca2d54772b8960d7d868fb75f8936e/tests/python/relay/utils/external_codegen.py#L76

Reply:
Thanks for this remark! I am a little bit confused about things regarding activating/deactivating a pass at the moment.
If I understand well, there is two ways to disable a pass :

1) In driver_api.cc, we have the registration of options:
```
// Register build pipeline related options
[...]
TVM_REGISTER_PASS_CONFIG_OPTION("tir.disable_vectorize", Bool);
TVM_REGISTER_PASS_CONFIG_OPTION("tir.disable_storage_rewrite", Bool);
TVM_REGISTER_PASS_CONFIG_OPTION("tir.is_entry_func", Bool);
TVM_REGISTER_PASS_CONFIG_OPTION("tir.add_lower_pass", Array<Array<ObjectRef>>);
TVM_REGISTER_PASS_CONFIG_OPTION("tir.debug_keep_trivial_loop", Bool);
```

Then in Array<tvm::transform::Pass> CreatePassList(), we get these booleans from the `config`:
```
  bool disable_vectorize = pass_ctx->GetConfig<Bool>("tir.disable_vectorize", Bool(false)).value();
  bool disable_storage_rewrite =
      pass_ctx->GetConfig<Bool>("tir.disable_storage_rewrite", Bool(false)).value();
  bool instrument_bound_checkers =
      pass_ctx->GetConfig<Bool>("tir.instrument_bound_checkers", Bool(false)).value();
  bool disable_cse_tir = pass_ctx->GetConfig<Bool>("tir.disable_cse_tir", Bool(false)).value();
```

And finally when we push the passes into the `pass_list`, we use the boolean that we got.
For instance:
```
[...]
 pass_list.push_back(tir::transform::VectorizeLoop(!disable_vectorize));
 pass_list.push_back(tir::transform::CommonSubexprElimTIR(!disable_cse_tir));
```

That makes sense to me.
Regarding what you said about the list of disabled pass, I assume that when a pass is mentioned in the disabled pass list, the context will get this information (somehow), and so the information will later flow through the GetConfig() showed above.
Am I correct on this?

2) In addition, some passes, like Vectorize, expose a boolean to the python interface for activating/deactivating them.
In transform.py:
```
def VectorizeLoop(enable_vectorize: bool = True):
    """Lower vectorization loops.
    [...]
    return _ffi_api.VectorizeLoop(enable_vectorize)  # type: ignore
```

I decided to do the same for the CSE. I though that would make it easier to disable the pass from the Python bindings, for instance for doing some experiments. If I remember well I was told that exposing the boolean would be useful (perhaps for the end-to-end compilation?).
Why is it done for Vectorize exactly? Should I keep it for the CSE?

Side note :
The pass LoopPartition is quite weird, as its boolean for activation/deactivation is passed directly to the `CreatePassList()` method:
`Array<tvm::transform::Pass> CreatePassList(bool disable_loop_partition)`
Why doesn't it get its boolean from the config, like all the other passes do?

Any precision would be very much appreciated :) Thanks!

Reply:
I didn't know that other TIR passes have a specific option for disabling. `disabled_pass` I brought up is enforced at https://github.com/apache/tvm/blob/3fbce70a8ad7e032de8c402fb27e0396435c8eca/src/ir/transform.cc#L479. I believe this has higher precedence than other disabling options, since it is enforced at the higher level in the stack. I don't know why things are like this way, maybe @tqchen @zhiics knows?

For Vectorize, I think this option is needed because we need to turn a loop with `vectorized` annotation into a sequential one if we want to skip this pass. 

https://github.com/apache/tvm/blob/c6f62aafc91e2600ed7772597fd4238c924c2a1b/src/tir/transforms/vectorize_loop.cc#L584-L588 

Not so long ago, the contents of `driver_api.c`` was actually written in python, so exposing this option to python made sense at that time. It's still good to have it today for experimentation purposes like you said. You can keep it for CSE if you find it useful, but for the purpose of disabling CSE, `disabled_list` already does the job (but not for `Vectorize` as explained above).

> The pass LoopPartition is quite weird, as its boolean for activation/deactivation is passed directly to the CreatePassList() method:

Indeed it is weird, I remember our `LoopPartition` implementation was problematic in the past (may be it is still today), probably an artifact from that time. Anyway I don't think there is a deep reason behind it.

Reply:
Many thanks for the very detailed @masahi, I really appreciate your comments and your time.
Ok, so I will keep for now the boolean to disable the pass easily, and we can still remove it in the near future if people really don't want these anymore.

I'm almost ready to send a final commit for this PR.
Many thanks again!